frontend/lib/js/category-trees/reducer.js

```diff
@@ -10,24 +10,113 @@ import {
   LOAD_TREE_SUCCESS,
   LOAD_TREE_ERROR,
   CLEAR_TREES,
-}                       from './actionTypes';
+  SEARCH_TREES_START,
+  SEARCH_TREES_END,
+}                                         from './actionTypes';
 
-import { setTree }      from './globalTreeStoreHelper';
+import { setTree, getConceptById }        from './globalTreeStoreHelper';
 
 export type TreesType = { [treeId: string]: NodeType }
 
+export type SearchType = {
+  searching: boolean,
+  searchStr: string,
+  words: Array<string>,
+  result: TreesType
+}
+
 export type StateType = {
   loading: boolean,
   version: any,
-  trees: TreesType
+  trees: TreesType,
+  search?: SearchType,
 };
 
 const initialState: StateType = {
   loading: false,
   version: null,
-  trees: {}
+  trees: {},
+  search: { searching: false, words: [], result: {} }
 };
 
+const searchTreesEnd = (state: StateType, action: Object): StateType => {
+  return {
+    ...state
+  }
+}
+
+const searchTreesStart = (state: StateType, action: Object): StateType => {
+  const { searchStr } = action.payload;
+  const searching = searchStr.length > 2;
+
+return {
+    ...state,
+    search: {
+      searching: searching,
+      searchStr: searchStr,
+      words: searchStr ? searchStr.split(' ') : [],
+      result: searching ? searchingTrees(state.trees, searchStr) : {}
+    }
+  }
+}
+
+const searchingTrees = (trees: TreesType, searchStr) => {
+  const regex = new RegExp(searchStr
+    .replace(/[\\[\]\\{}()+*?.$^|]/g,
+    function (match) { return '\\' + match }), "i");
+
+  return Object.assign({}, ...Object.entries(trees).map(([treeId, treeNode]) => ({
+    [treeId]: findTreeNodes(treeId, treeNode, regex)
+  })).filter(r => { return Object.keys(r).some(k => r[k].length > 0) }));
+}
+
+const findTreeNodes = (treeId: string, treeNode: NodeType, searchStr: string) => {
+  if (treeNode === null)
+    treeNode = getConceptById(treeId);
```

Comment:
A couple of thoughts here:

 - Maybe we can avoid re-assigning method parameters
 - I've had a quick look at the CPU profiler: getConceptById takes up about 70% of the total search time. I'm not sure if we can improve on that in general, but the specific implementation doesn't seem optimal. Maybe we can have a variant of that method where we pass in the concept tree id as a parameter, rather than letting the method itself search all trees for the provided id.
 - Let's be consistent with naming (here and in general): When something is called `tree` or `treeId` it should always refer to the root node of a concept tree (i.e. the one with `tables` set, **not** a structure node and **not** any other node of a concept tree)